tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,76 @@
+#ifndef ROOT_RDF_RMERGEABLEVALUE
+#define ROOT_RDF_RMERGEABLEVALUE
+
+#include <functional>
+#include <memory>
+#include "RtypesCore.h"
+
+namespace ROOT {
+namespace RDF {
+
+template <typename T>
+class RMergeableValue {
+protected:
+   T fValue;
+
+public:
+   RMergeableValue() = delete;
+   RMergeableValue(T &value) : fValue{value} {}
+   virtual ~RMergeableValue() = default;
+   virtual void Merge(const std::unique_ptr<RMergeableValue<T>> &) = 0;
+   virtual ULong64_t GetCounts() { return 0; }
+   virtual T &GetValue() { return fValue; }
+};
+
+// Specialization for generic action
+template <typename T>
+class RMergeableGenericAction final : public RMergeableValue<T> {
+   using mergefun_t = std::function<T(T &, T &)>;
```

Comment:
This signature doesn't fit with some `T`s, e.g. `TH1D`, since the merging in that case would be done on the (left) object itself
```
TH1D Merge(TH1D& left, TH1D& right){
left.Merge(&right);
return left
}
```
Although I have already showed this signature to be a bit flawed since it involves an additional copy, see my comments on the addition of `RMergeableHisto1D`

Reply:
It has to be one of these two:

* `T(const T&, const T&)`: you don't modify the arguments, but create a third object instead, or
* `void(T&, const T&)`: you accumulate into the first argument, don't need a return value

`T(T&, T&)` can do too much.

Reply:
Ok I'll go with `T(const T&, const T&)` that looks more generic. I will address other signatures as needed according to the action they're dealing with